Thermostat presets

MrMDavidson opened the pull request:
# What does this implement/fix?

This introduces the ability to configure additional presets for a `Thermostat` beyond the existing Home and Away modes. The existing configuration will continue to work. But it also introduces a new configuration list for preset modes. See the example config.

The existing Home/Away logic has been consolidated with this generic preset approach. I'd be keen to further extend this to have the presets also configure the default operation mode/fan mode/swing mode as well as just the temperature.

Apologies for any faux pas in the Python side of it - it's not my forte.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

(Note: Yet to update the docs, wanted to see if this is something of interest to the community/the config stays stable before updating them)

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
climate:
  - platform: thermostat
    # Kept for backwards compatability
    away_config:
      default_target_temperature_low: 16
      default_target_temperature_high: 26

    preset:
      activity:
        default_target_temperature_low: 21
        default_target_temperature_high: 23
      sleep:
        default_target_temperature_low: 17
        default_target_temperature_high: 26
      # etc. Can be any of eco, away, boost, comfort, home, sleep, and activity
    preset_change:
      - logger.log: Preset has been changed
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

MrMDavidson committed:
Rework HOME/AWAY support to being driven via a map of ClimatePreset/ThermostatClimateTargetTempConfig
This opens up to theoretically being able to support other presets (ECO, SLEEP, etc)

MrMDavidson committed:
Rework HOME/AWAY support to being driven via a map of ClimatePreset/ThermostatClimateTargetTempConfig
This opens up to theoretically being able to support other presets (ECO, SLEEP, etc)

MrMDavidson committed:
Add support for additional presets
Configuration takes the form;
```
climate:
  platform: preset
  ...
  preset:
    [eco | away | boost | comfort | home | sleep | activity]:
      default_target_temperature_low: 20
      default_target_temperature_high: 24
```

These will be available in the Home Assistant UI and, like the existing Home/Away config will reset the temperature in line with these defaults when selected. The existing away_config/home_config is still respected (although preset->home/preset->away will be applied after them and override them if both styles are specified)

MrMDavidson committed:
Add support for additional presets
Configuration takes the form;
```
climate:
  platform: preset
  ...
  preset:
    [eco | away | boost | comfort | home | sleep | activity]:
      default_target_temperature_low: 20
      default_target_temperature_high: 24
```

These will be available in the Home Assistant UI and, like the existing Home/Away config will reset the temperature in line with these defaults when selected. The existing away_config/home_config is still respected (although preset->home/preset->away will be applied after them and override them if both styles are specified)

MrMDavidson committed:
Remove trailing whitespace

MrMDavidson committed:
Remove trailing whitespace

MrMDavidson committed:
Apply clang suggestions

MrMDavidson committed:
Apply clang suggestions

MrMDavidson committed:
Apply lint-python suggestions

MrMDavidson committed:
Apply lint-python suggestions

MrMDavidson committed:
Fix compilation issues from prior formatting tweaks

MrMDavidson committed:
Fix compilation issues from prior formatting tweaks

MrMDavidson committed:
More style tweaks

MrMDavidson committed:
More style tweaks

MrMDavidson committed:
Apply results of clang-tidy

MrMDavidson committed:
Apply results of clang-tidy

MrMDavidson committed:
Run climate.py through Black formatter

MrMDavidson committed:
Run climate.py through Black formatter

MrMDavidson committed:
Additional clang-tidy fixes

MrMDavidson committed:
Additional clang-tidy fixes

nagyrobi commented:
This is excellent, and we need this for `climate`, `bang_bang`, `pid` climate platforms too.

kbx81 commented:
Hey! Thanks for working on this! I think it will make a lot of people happy.

I had a cursory look over the changes and I think it looks good -- I'll try to get around to testing and playing around with it more over the next week or two. 😄

MrMDavidson commented:
> Hey! Thanks for working on this! I think it will make a lot of people happy.

No worries - it was a bit of a gap in my personal setup so figured I'd have a crack at implementing it. I think the configuration validation will need some tweaks / has some gaps, so happy for any feedback in that area.
 
> I had a cursory look over the changes and I think it looks good -- I'll try to get around to testing and playing around with it more over the next week or two. 😄

Do you think it's worth expanding the functionality over what a preset covers? I'd personally be keen to see the fan mode as part of the preset (Eg. Sleep mode also sets the fan to low) but if I'm doing one it's not much more effort to expand it to others. It does raise questions around if the user changes both the preset and a mode controlled by the preset which takes priority (I'd say the manually specified one?) Eg. Using my example of having the Sleep 
```yaml
  on...:
    then:
      - climate.control:
          id: my_aircon
          preset: sleep
          fan_mode: HIGH
```

wilberforce committed:
Webserver v2 (#2688)

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

JasperPlant committed:
Add entity_category_diagnostics to SGP30 baseline sensors (#3272)

MrMDavidson commented:
> Haha, well, full disclosure: I don't have strong feelings about presets -- I just don't use them. 😇 I think that, traditionally, presets just define set points, but I don't see anything specifically _wrong_ with allowing the mode/fan mode/swing mode/etc. to be adjusted with them, as well. But that's just my two cents...others here might have different feelings on the matter. 🙃

Ah, in my case I'm having my controller change presets in response to events. Eg. In the room? Switch to "Activity" preset with a tighter temperature range vs just at home but not in the room vs away from home.

I've actually got another branch with support for controlling these extra modes as part of the presets and it's working well. Will keep playing with it (need to test when they've got a lambda, for instance) and if it works well and there's no opposition I'll merge it into this branch.

nagyrobi commented:
Can the preset temp values be set at run time? Say, with lambdas pointing at input_numbers?

MrMDavidson committed:
Add support for specifying MODE, FAN_MDOE and SWING_MODE on a preset
When switching presets these will implicitly flow through to the controller. However calls to climate.control which specify any of these will take precedence even when changing the mode (think of the preset version as the default for that preset)

MrMDavidson committed:
Add support for specifying MODE, FAN_MDOE and SWING_MODE on a preset
When switching presets these will implicitly flow through to the controller. However calls to climate.control which specify any of these will take precedence even when changing the mode (think of the preset version as the default for that preset)

MrMDavidson commented:
> Can the preset temp values be set at run time? Say, with lambdas pointing at input_numbers?

I had a quick play but I'm not sure if this is possible overloading the same configuration value for both a compile time value and a run time value. Happy to be corrected, here? But having a look through the docs/code I can't find examples for non-actions that do this. So as far as I can tell this would only be possible if we introduced a second configuration type for when you need it configured via lambda. Something like;
```yaml
    preset:
      activity:
        default_target_temperature_low_templatable: !lambda |-
          return id(my_home_assistant_exported_value).state;
        default_target_temperature_high: 23
```

I think I had a very similar setup to what you're describing and, honestly, it got very messy on the front end and was the driving reason to implement support for presets instead. I can sort of see the use case for lambdas here to return a dynamic value based on something a sensor knows (eg. More than 2 people in the room? Bump the set temperature set points down an extra degree to keep the room cool, or something). That said, these would only be evaluated at switching time (Continuing this example; "There's two people in the room _now_ we're switching to 'Activity' mode" but if another person entered, it wouldn't react to it). With this in mind I think it might be better to leave this functionality out and instead make use of triggers elsewhere in your configuration to prod the state of the A/C? For your use case you can still do something like;

```yaml
  on...
    then:
      - climate.control:
          id: my_aircon
          preset: CLIMATE_PRESET_HOME
          target_temperature_low: !lambda |-
            return id(my_front_end_driven_value_low).state;
          target_temperature_high: !lambda |-
            return id(my_front_end_driven_value_high).state;
```

nagyrobi commented:
The use case is much more simpler: implementing a control user interface where the various temperatures desired for the presets can be set by the home inhabitants, without needing to recompile node firmware.

So one could configure input numbers (with some defaults, and restore from flash option set) for the various preset modes. The user is free to adjust these as they find it better. For example if you have such a thermostat for each room, it may very well be required that different room inhabitants feel comfortable at different activity or sleep temperatures so they would need to be given the opportunity to customize these values themselves.

jesserockz committed:
Bump esphome-dashboard to 20220309.0 (#3277)

wilberforce committed:
Add visual step/min/max for webserver climate (#3275)

jesserockz committed:
Bump version to 2022.4.0-dev

dependabot[bot] committed:
Bump pytest-asyncio from 0.18.1 to 0.18.2 (#3262)

Bumps [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio) from 0.18.1 to 0.18.2.
- [Release notes](https://github.com/pytest-dev/pytest-asyncio/releases)
- [Commits](https://github.com/pytest-dev/pytest-asyncio/compare/v0.18.1...v0.18.2)

---
updated-dependencies:
- dependency-name: pytest-asyncio
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump zeroconf from 0.38.3 to 0.38.4 (#3257)

Bumps [zeroconf](https://github.com/jstasiak/python-zeroconf) from 0.38.3 to 0.38.4.
- [Release notes](https://github.com/jstasiak/python-zeroconf/releases)
- [Commits](https://github.com/jstasiak/python-zeroconf/compare/0.38.3...0.38.4)

---
updated-dependencies:
- dependency-name: zeroconf
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump click from 8.0.3 to 8.0.4 (#3248)

Bumps [click](https://github.com/pallets/click) from 8.0.3 to 8.0.4.
- [Release notes](https://github.com/pallets/click/releases)
- [Changelog](https://github.com/pallets/click/blob/main/CHANGES.rst)
- [Commits](https://github.com/pallets/click/compare/8.0.3...8.0.4)

---
updated-dependencies:
- dependency-name: click
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

OttoWinter committed:
Bump docker dependencies (#3281)

stegm committed:
Add optimistic config flag to modbus select. (#3267)

Rai-Rai committed:
Fixed wrong comment (#3286)

MrMDavidson committed:
Add `preset_change` mode trigger
When defined this trigger will fire when the preset for the thermostat has been changed. The intent of this is similar to `auto_mode` - it's not intended to be used to control the preset's state (eg. communicate with the physical thermostat) but instead might be used to update a visual indicator, for instance.

MrMDavidson committed:
Add `preset_change` mode trigger
When defined this trigger will fire when the preset for the thermostat has been changed. The intent of this is similar to `auto_mode` - it's not intended to be used to control the preset's state (eg. communicate with the physical thermostat) but instead might be used to update a visual indicator, for instance.

kbx81 commented on esphome/components/thermostat/climate.py:
```diff
@@ -58,8 +58,14 @@
+    CONF_MODE
 )
 
+CONF_PRESET_CHANGE = "preset_change"
```
Neat!...but I think if you're going to build a "preset change" trigger, it should be either "on_preset" or "on_preset_change" so it's consistent with other triggers as outlined in the [docs](https://next.esphome.io/guides/automations.html#all-triggers). 😄

MrMDavidson replied:
There's definitely some variance on the naming conventions here but was trying to vaguely make it align with the existing ones for the thermostat (eg. `CONG_AUTO_MODE = "auto_mode"`). My preference would be to make them past tense (`on_preset_changed`) but happy to take your input on preferred naming convention here. Preferences?

kbx81 replied:
Ah! I'm reading things too fast again. You're adding an action, not a trigger. Not sure what I was thinking...brain went to something else.

I think `preset_change` is fine. Nothing to see here, move along. 😆

kbx81 commented on esphome/components/thermostat/climate.py:
```diff
@@ -183,7 +202,6 @@ def validate_thermostat(config):
                 raise cv.Invalid(
                     f"{req_trigger} must be defined to use {config_trigger}"
                 )
```
Total nitpick but...is it really necessary to remove all the blank lines? Visually, for me, they help create separation between blocks of code operating on logically different things. It helps somewhat with readability (I'm a big fan of readability 😄). It also just makes the PR bigger. 😇

MrMDavidson replied:
I'm more than happy to leave them - I only stripped them as the build failures suggesting running the code through the Black formatter. The one I happened to be using (online) stripped them out

kbx81 replied:
Oh dear -- I hope our linter/formatter configuration isn't stripping them. I'd be really sad. Are you using VSCode? There is an extension so I think you can just do it locally...and I think it'll use our preferred configuration. 🙂

MrMDavidson replied:
Alright, let me get that set up and re-lint the files.

kbx81 reviewed (commented):
One legit concern, one nitpick. Nothing big. 🙂

MrMDavidson committed:
Fix formatting

MrMDavidson committed:
Fix formatting

MrMDavidson committed:
Apply Clang suggestions

MrMDavidson committed:
Merge branch 'thermostat-presets' of github.com:MrMDavidson/esphome into thermostat-presets

MrMDavidson commented:
Apologies, git rebase gone awry. It's late at night here I'll abandon this and re-open it tomorrow without the mess.

kbx81 commented:
Alrighty! No worries -- if you can fix your branch you can force-push it and it'll fix up the "mess". 😄
If you make a new PR, add a comment and mention this one so we can easily find it. 🙂
